core/domain/classifier/base_classification_algorithm.py

```diff
@@ -0,0 +1,153 @@
+# coding: utf-8
+#
+# Copyright 2015 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Base class of classification algorithms for free-form text answers"""
+
+from abc import abstractmethod
+import copy
+
+
+class BaseClassificationAlgorithm(object):
+    """A base class for classifiers that uses supervised learning to match
```

Comment:
Oh, I thought we only focus on supervised learning on text-format examples and the descendant classes are only for different models such as SVM, LDA and etc, so the input examples should be in the same format for all concrete implementations.

Do we really need to handle the input examples of other formats? If so, we should have another layer to abstract the common behaviors for a certain input type. For instance, BaseClassificationAlgorithm -> StringClassifier -> LDAStringClassifier.